File: tools/t_class_classifier/cpp/classify.cpp
#include "schemas.hpp"

ClassificationResult assign_t_class(const TSig& sig) {
    ClassificationResult res;
    res.t_sig = sig;

    int C_count = sig.C_count;
    int B_cross = sig.B_cross;
    int component_count = sig.component_count;
    int raw_edge_count = sig.raw_edge_count;
    int unique_edge_count = sig.unique_edge_count;
    int parallel_edge_count = sig.parallel_edge_count;

    if (C_count == 0) {
        if (unique_edge_count > 0 || raw_edge_count > 0 || parallel_edge_count > 0) {
            res.t_class = "T_x";
            res.is_valid_closure = true;
        } else {
            res.t_class = "T_0";
            res.is_valid_closure = false;
        }
    } else if (C_count == 1) {
        res.t_class = "T_1";
        res.is_valid_closure = true;
    } else if (C_count == 2 && (component_count >= 2 || B_cross <= 2)) {
        res.t_class = "T_2";
        res.is_valid_closure = true;
    } else if (C_count >= 3) {
        if (B_cross <= 3) {
            res.t_class = "T_3";
        } else {
            res.t_class = "T_4";
        }
        res.is_valid_closure = true;
    } else {
        res.t_class = "T_x";
        res.is_valid_closure = true;
    }

    return res;
}

File: tools/t_class_classifier/python/classify.py
from schemas import TSig, ClassificationResult

def assign_t_class(t_sig: TSig) -> ClassificationResult:
    C_count = t_sig.C_count
    B_cross = t_sig.B_cross
    component_count = t_sig.component_count
    raw_edge_count = t_sig.raw_edge_count
    unique_edge_count = t_sig.unique_edge_count
    parallel_edge_count = t_sig.parallel_edge_count

    if C_count == 0:
        if unique_edge_count > 0 or raw_edge_count > 0 or parallel_edge_count > 0:
            t_class = "T_x"
            is_valid_closure = True
        else:
            t_class = "T_0"
            is_valid_closure = False
    elif C_count == 1:
        t_class = "T_1"
        is_valid_closure = True
    elif C_count == 2 and (component_count >= 2 or B_cross <= 2):
        t_class = "T_2"
        is_valid_closure = True
    elif C_count >= 3:
        if B_cross <= 3:
            t_class = "T_3"
        else:
            t_class = "T_4"
        is_valid_closure = True
    else:
        t_class = "T_x"
        is_valid_closure = True

    return ClassificationResult(
        t_sig=t_sig,
        t_class=t_class,
        is_valid_closure=is_valid_closure
    )
